### DISCUSSION TITLE
About the orientation in cartesian space control

### DISCUSSION THREAD

**Discussion by Yujia-Zhang0913** [robotology/community#665] [Q&A] [answered]
Hi,
I am running the code of cartesian space control and my code is like this:

```cpp
Vector position_vector, orientation_vector;
 while (!icart->getPose(position_vector, orientation_vector))
     Time::yield();  // remember to release the quantum
 Vector position_vector1 = position_vector; 
 position_vector1[2]+=0.1;
 Vector orientation_vector1=orientation_vector;
 //Wait motion done
 icart->goToPoseSync(position_vector1, orientation_vector1);
 yInfo() << "Waiting motion done...";
 icart->waitMotionDone(0.04, 3.0); //waiting for 3 seconds max
```

However, I don't know why when I changed the pose to make icub move along x or y or z, the orientation of the hand is different. I think since I didn't change the orientation, it should be the same.

- move along z
![2023-11-04 11-13-54 的屏幕截图](https://github.com/robotology/community/assets/73634487/fe157cc3-67d5-45e7-9817-23242320aa7c)

- move along y
![2023-11-04 11-16-38 的屏幕截图](https://github.com/robotology/community/assets/73634487/7334b83c-bc0a-476f-af9e-a77b4076e5c4)

- move along x
![2023-11-04 11-17-50 的屏幕截图](https://github.com/robotology/community/assets/73634487/caf1ba6a-c258-4e61-8987-a94678fea256)

And in the tutorial[https://robotology.github.io/robotology-documentation/doc/html/icub_cartesian_interface.html](url), if I write code like this:

```cpp
Vector oy(4);
 oy[0]=0.0; oy[1]=1.0; oy[2]=0.0; oy[3]=+M_PI/2.0;    
 Matrix Ry=yarp::math::axis2dcm(oy);        // from axis/angle to rotation matrix notation
 Matrix R0=yarp::math::axis2dcm(orientation_vector);
    
  Matrix R=R0*Ry;                            // compose the two rotations keeping the order
  Vector o=yarp::math::dcm2axis(R);          // from rotation matrix back to the axis/angle notation
```

Is this meaning that the hand will rotate of pi/2 around the y-axis of current orientation?
I have troubling in the orientation change. Really thanks.

**Comment by pattacini** [comment, marked as answer]
Hi @Yujia-Zhang0913 

Try to enable the torso by means of the [setDOF()](https://yarp.it/latest/classyarp_1_1dev_1_1ICartesianControl.html#a3198392b8807c1c293982301b658cc9c) service. Just go with a whole 10-DOF reaching. Find more on this at https://robotology.github.io/robotology-documentation/doc/html/icub_cartesian_interface.html#sec_cart_useinterface.

The iCub has a narrow dexterous space when using only the 7-DOF of the arm. Also, reaching in position has a higher priority wrt reaching in orientation. This has a particular impact when targets are barely reachable given the current settings. Don't know if this is the case for your experiments, but it's definitely worth giving it a try.

>Is this meaning that the hand will rotate of pi/2 around the y-axis of current orientation?

Yep 👍🏻 
For example, the [shrug move 📽️][shrug-movie] is implemented by this [function][shrug-code], where the left orientation is derived from the right one with a similar rotation.

[shrug-movie]: https://robotology.github.io/icub-gazebo-grasping-sandbox/assets/videos/shrug.mp4
[shrug-code]: https://github.com/robotology/icub-gazebo-grasping-sandbox/blob/c9e5edd8c40b6af602dfdfafa41a16516ca2383a/src/main.cpp#L126C22-L144

**Reply by Yujia-Zhang0913** [reply]
Yes! It works out when I enable the torso. Thanks a lot!

**Reply by pattacini** [reply]
Happy it worked 👍🏻
